app/models/plan.rb

```diff
@@ -26,7 +27,7 @@ class PeriodRangeCalculationError < StandardError; end
   #  - `#audit_destroy` triggered by audit_destroy
   # I think those gems are fine with their behaviour.
   # The problem is on our side with so many inter-dependent relationships
-  skip_callback :destroy, :before, :lock!, if: -> { !self.class.exists?(id) || destroyed_by_association }
+  # skip_callback :destroy, :before, :lock!, if: -> { !self.class.exists?(id) || destroyed_by_association }
```

Comment:
I am not sure if this is entirely correct. Although it seems so, but I prefer to express my concerns to be sure :smile: 

Part of this has a test in master:
https://github.com/3scale/porta/blob/e2dbb0cf87ed7a340261f681b9444254dd1a199b/test/unit/plan_test.rb#L603-L615
And this is passing :heavy_check_mark: :+1: 

And locally I added a test for the part that was untested there (I am not sure if it is tested somewhere else, but I think not. It was added in https://github.com/3scale/system/pull/9244) :
```ruby
def test_plan_not_locked_if_destroyed_by_association
  plan = FactoryBot.create(:simple_application_plan)
  plan.class.stubs(:exists?).with(plan.id).returns(true)
  plan.class.stubs(:destroyed_by_association).with(plan.id).returns(true)
  plan.expects(:lock!).never
  plan.destroy
end
```

And this also passes when I do `bundle exec rake test TEST=test/unit/plan_test.rb`
So my conclusions are:
1. We should add this test, because it is a scenario that matters.
2. Does this test run callbacks? because otherwise, this is passing for the wrong reasons :joy:

Reply:
All callbacks since rails 5.0 are called in tests

Reply:
By default yes, but sometimes we change that for some tests :smile: But I see that it isn't the case :+1: Good :+1: I will add the tests in a separate PR.

Reply:
https://github.com/3scale/porta/pull/1666 :smile: